src/routes/OAuth2.mjs

```diff
@@ -1,269 +1,460 @@
 import API, {
-  clientAuthenticated,
   authenticated,
+  clientAuthenticated,
   GET,
-  POST,
-  required,
   parameters,
+  POST,
 } from './API'
 
+const transactionTimeoutMinutes = 10
+const transactionTimeout = transactionTimeoutMinutes * 60 * 1000
 const sessionExpiryTime = 60 * 60 * 1000
 
-const server = oauth2orize.createServer()
-
-server.serializeClient((client) => {
-  return client.id
-})
+/**
+ * Class for managing OAuth 2 requests
+ * @class
+ */
+class OAuth extends API {
+  static transactions = new Map()
 
-server.deserializeClient(async (id) => {
-  const client = await Client.scope('user').findByPk(id)
-  if (!client) {
-    return false
+  /**
+   * @inheritdoc
+   */
+  get type () {
+    return 'oauth2'
   }
 
-  return client
-})
-
-server.grant(oauth2orize.grant.code(async (client, redirectUri, user, ares, areq) => {
-  Permission.assertOAuthScopes(areq.scope)
-
-  const clientRedirectUri = redirectUri ?? client.redirectUri
-
-  const code = await Code.create({
-    value: await oAuthTokenGenerator(),
-    scope: areq.scope,
-    redirectUri: clientRedirectUri,
-    clientId: client.id,
-    userId: user.id,
-  })
-  return code.value
-}))
-
-server.grant(oauth2orize.grant.token(async (client, user, ares, areq) => {
-  Permission.assertOAuthScopes(areq.scope)
-
-  const token = await Token.create({
-    value: await oAuthTokenGenerator(),
-    scope: areq.scope,
-    clientId: client.id,
-    userId: user.id,
-  })
-  return token.value
-}))
-
-server.exchange(oauth2orize.exchange.code(async (client, code, redirectUri) => {
-  const auth = await Code.findOne({ where: { value: code } })
-
-  if (!auth || client.id !== auth.clientId || redirectUri !== auth.redirectUri) {
-    return false
-  }
+  /**
+   * Endpoint for OAuth authorize decision screen info requests
+   * @endpoint
+   */
+  @GET('/oauth2/authorize')
+  @authenticated
+  @parameters('client_id')
+  async authorize (ctx) {
+    const {
+      response_type: responseType,
+      client_id: clientId,
+      redirect_uri: redirectUri,
+      scope,
+      state,
+    } = ctx.query
+
+    if (!isValidRedirectUri(redirectUri)) {
+      throw new BadRequestAPIError({ parameter: 'redirect_uri' })
+    }
 
-  await auth.destroy()
+    if (!state) {
+      return callbackError(redirectUri, new InvalidRequestOAuthError('state'))
+    }
 
-  const token = await Token.create({
-    scope: auth.scope,
-    value: await oAuthTokenGenerator(),
-    clientId: client.id,
-    userId: auth.userId,
-  })
-  return token.value
-}))
+    if (!responseType) {
+      return callbackError(redirectUri, new InvalidRequestOAuthError('response_type'))
+    }
 
-server.exchange(oauth2orize.exchange.password(async (client, username, password, scope, ctx) => {
-  if (!client) {
-    return false
-  }
-  if (!client.firstParty) {
-    throw new ForbiddenAPIError({ parameter: 'username' })
-  }
+    if (!scope || typeof scope !== 'string' || scope.length === 0) {
```

Comment:
as above, `!scope` is redundant if explicit `typeof` and `.length` checks are made.